From e786f555d397e24b394965c2ea6d1e158d83949e Mon Sep 17 00:00:00 2001
From: Han Xu <han.xu@nxp.com>
Date: Wed, 8 Nov 2017 15:55:27 -0600
Subject: [PATCH 2754/5242] MLK-16786: dma: mxs-dma: fix the unbalanced
 runtime pm counter

commit  8ac8243ce87135e3110b53176c050e73f5017855 from
https://source.codeaurora.org/external/imx/linux-imx.git

mxs-dma init function will call runtime pm init, the redundant
pm_runtime_force_resume will mess up the counter. Also remove some
unnecessary code.

Signed-off-by: Han Xu <han.xu@nxp.com>
Reviewed-by: Frank Li <frank.li@nxp.com>

During 4.14 rebase squashed the following:

MLK-16800: dma: mxs-dma: correctly handle mxs-dma clock

enable mxs-dma clock before HW reset and disable clock after it.

Signed-off-by: Leonard Crestez <leonard.crestez@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/dma/mxs-dma.c |   10 ++++------
 1 file changed, 4 insertions(+), 6 deletions(-)

diff --git a/drivers/dma/mxs-dma.c b/drivers/dma/mxs-dma.c
index e18f7d9..d374548 100644
--- a/drivers/dma/mxs-dma.c
+++ b/drivers/dma/mxs-dma.c
@@ -1,6 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0
 //
 // Copyright 2011-2015 Freescale Semiconductor, Inc. All Rights Reserved.
+// Copyright 2017 NXP
 //
 // Refer to drivers/dma/imx-sdma.c
 
@@ -724,8 +725,8 @@ static int mxs_dma_init_rpm(struct mxs_dma_engine *mxs_dma)
 
 static int mxs_dma_init(struct mxs_dma_engine *mxs_dma)
 {
-	int ret;
 	struct device *dev = &mxs_dma->pdev->dev;
+	int ret;
 
 	ret = mxs_dma_init_rpm(mxs_dma);
 	if (ret)
@@ -753,10 +754,10 @@ static int mxs_dma_init(struct mxs_dma_engine *mxs_dma)
 	writel(MXS_DMA_CHANNELS_MASK << MXS_DMA_CHANNELS,
 		mxs_dma->base + HW_APBHX_CTRL1 + STMP_OFFSET_REG_SET);
 
-err_clk:
 	pm_runtime_mark_last_busy(dev);
 	pm_runtime_put_autosuspend(dev);
 
+err_clk:
 	return ret;
 }
 
@@ -930,13 +931,10 @@ static int mxs_dma_pm_resume(struct device *dev)
 	struct mxs_dma_engine *mxs_dma = dev_get_drvdata(dev);
 	int ret;
 
-	ret = pm_runtime_force_resume(dev);
-	if (ret)
-		return ret;
-
 	ret = mxs_dma_init(mxs_dma);
 	if (ret)
 		return ret;
+
 	return 0;
 }
 
-- 
1.7.9.5

